nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarProducerProcessor.java

```diff
@@ -0,0 +1,323 @@
+            .required(false)
+            .allowableValues(MESSAGE_ROUTING_MODE_CUSTOM_PARTITION, MESSAGE_ROUTING_MODE_ROUND_ROBIN_PARTITION, MESSAGE_ROUTING_MODE_SINGLE_PARTITION)
+            .defaultValue(MESSAGE_ROUTING_MODE_ROUND_ROBIN_PARTITION.getValue())
+            .build();
+
+    public static final PropertyDescriptor PENDING_MAX_MESSAGES = new PropertyDescriptor.Builder()
+            .name("Max Pending Messages")
+            .description("Set the max size of the queue holding the messages pending to receive an "
+                    + "acknowledgment from the broker.")
+            .required(false)
+            .addValidator(StandardValidators.POSITIVE_INTEGER_VALIDATOR)
+            .defaultValue("1000")
+            .build();
+
+    protected static final List<PropertyDescriptor> PROPERTIES;
+    protected static final Set<Relationship> RELATIONSHIPS;
+
+    static {
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.add(PULSAR_CLIENT_SERVICE);
+        properties.add(TOPIC);
+        properties.add(ASYNC_ENABLED);
+        properties.add(MAX_ASYNC_REQUESTS);
+        properties.add(BATCHING_ENABLED);
+        properties.add(BATCHING_MAX_MESSAGES);
+        properties.add(BATCH_INTERVAL);
+        properties.add(BLOCK_IF_QUEUE_FULL);
+        properties.add(COMPRESSION_TYPE);
+        properties.add(MESSAGE_ROUTING_MODE);
+        properties.add(PENDING_MAX_MESSAGES);
+        PROPERTIES = Collections.unmodifiableList(properties);
+
+        final Set<Relationship> relationships = new HashSet<>();
+        relationships.add(REL_SUCCESS);
+        relationships.add(REL_FAILURE);
+        RELATIONSHIPS = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    protected LRUCache<String, Producer<T>> producers;
+
+    // Pool for running multiple publish Async requests
+    protected ExecutorService publisherPool;
+    protected ExecutorCompletionService<Object> publisherService;
+
+    @OnScheduled
+    public void init(ProcessContext context) {
+        if (context.getProperty(ASYNC_ENABLED).isSet() && context.getProperty(ASYNC_ENABLED).asBoolean()) {
+            publisherPool = Executors.newFixedThreadPool(context.getProperty(MAX_ASYNC_REQUESTS).asInteger());
+            publisherService = new ExecutorCompletionService<>(publisherPool);
+        }
+    }
+
+    @OnUnscheduled
+    public void shutDown(final ProcessContext context) {
+       if (context.getProperty(ASYNC_ENABLED).isSet() && context.getProperty(ASYNC_ENABLED).asBoolean()) {
+           // Stop all the async publishers
+           try {
+              publisherPool.shutdown();
+              publisherPool.awaitTermination(20, TimeUnit.SECONDS);
```

Comment:
We are waiting for the asynchronous message acknowledgments to return in order to prevent duplicate data being sent to Pulsar in the event of a shutdown and restart of the processor